models/issue.go

```diff
@@ -419,17 +433,56 @@ func (issue *Issue) ClearLabels(doer *User) (err error) {
 
 // ReplaceLabels removes all current labels and add new labels to the issue.
 // Triggers appropriate WebHooks, if any.
-func (issue *Issue) ReplaceLabels(labels []*Label) (err error) {
+func (issue *Issue) ReplaceLabels(labels []*Label, doer *User) (err error) {
 	sess := x.NewSession()
 	defer sessionRelease(sess)
 	if err = sess.Begin(); err != nil {
 		return err
 	}
 
-	if err = issue.clearLabels(sess); err != nil {
-		return fmt.Errorf("clearLabels: %v", err)
-	} else if err = issue.addLabels(sess, labels); err != nil {
-		return fmt.Errorf("addLabels: %v", err)
+	if err = issue.loadLabels(sess); err != nil {
+		return err
+	}
+
+	var toAdd, toRemove []*Label
+	for _, l := range labels {
```

Comment:
Maybe we should `Sort` these first? should make it slightly faster in some cases 🙂